File: app/src/main/java/com/konstant/tool/test/ForegroundService.java
package com.konstant.tool.test;

import android.app.Notification;
import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.Nullable;

/**
 * 描述:前台服务
 * 创建人:菜籽
 * 创建时间:2018/2/27 下午6:15
 * 备注:
 */

public class ForegroundService extends Service {

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    /**
     *
     * @param intent    保存传递过来的参数
     * @param flags
     * @param startId
     * @return
     */
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        createNotification();
        return super.onStartCommand(intent, flags, startId);
    }

    // 创建通知栏的通知
    public void createNotification(){
        // 创建状态栏通知的构造器
        Notification.Builder builder = new Notification.Builder(this);
        // 为构造器设置小图标
//        builder.setSmallIcon(R.drawable.ic_launcher);
//        // 为构造器设置大图标
//        builder.setLargeIcon(BitmapFactory.decodeResource(getResources(),R.drawable.ic_launcher));
//        // 禁止删除
//        builder.setAutoCancel(false);
//        // 右上角显示时间
//        builder.setShowWhen(true);
        // 设置通知栏标题
//        builder.setContentTitle("菜籽工具箱");
        // 创建通知
        Notification notification = builder.build();
        // 设置为前台通知，int：代表唯一标示通知的整形数；
        startForeground(1,notification);
//        NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
//        notificationManager.notify(1,notification);
    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        // 关闭前台通知，参数为true表示同时移除状态栏的通知
        stopForeground(true);
    }
}

File: tool/src/main/java/com/konstant/tool/lite/base/ForegroundService.kt
package com.konstant.tool.lite.base

import android.app.*
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import com.konstant.tool.lite.R
import com.konstant.tool.lite.module.setting.activity.SettingActivity

/**
 * 作者：konstant
 * 时间：2019/10/25 18:13
 * 描述：前台通知，用于加强应用存活性
 */

class ForegroundService : Service() {

    companion object {
        private const val ENABLE_ENHANCE = "enable_enhance"
        fun startForegroundService(context: Context, enhance: Boolean) {
            with(Intent(context, ForegroundService::class.java)) {
                putExtra(ENABLE_ENHANCE, enhance)
                context.startService(this)
            }
        }

        fun createForegroundNotification(context: Context, title: String = "菜籽工具箱-后台增强服务",
                                         msg: String = "关闭'后台增强服务'后，此通知会自动移除"): Notification {
            val channelId = "${context.packageName}.channel"
            val intent = TaskStackBuilder.create(context)
                    .addNextIntent(Intent(context, SettingActivity::class.java))
                    .getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT)
            val notificationBuilder = NotificationCompat.Builder(context, channelId)
                    .apply {
                        setSmallIcon(R.drawable.ic_launcher)
                        setLargeIcon(BitmapFactory.decodeResource(context.resources, R.drawable.ic_launcher))
                        setAutoCancel(false)
                        setShowWhen(true)
                        setContentTitle(title)
                        setContentText(msg)
                        setOngoing(true)
                        setContentIntent(intent)
                    }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val channelName = "后台增强服务"
                val channel = NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_MIN)
                val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
                manager.createNotificationChannel(channel)
            }
            return notificationBuilder.build()
        }
    }

    override fun onBind(intent: Intent): IBinder? = null

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val enhance = intent.getBooleanExtra(ENABLE_ENHANCE, false)
        if (enhance) {
            startForeground(1, Companion.createForegroundNotification(this))
        } else {
            stopForeground(true)
        }
        return super.onStartCommand(intent, flags, startId)
    }

}
